[SPARK-49829][SS] Fix the bug on the optimization on adding input to state store in stream-stream join

HeartSaVioR opened the pull request:
### What changes were proposed in this pull request?

The PR proposes to revise the optimization on adding input to state store in stream-stream join, to fix correctness issue.

### Why are the changes needed?

Here is the logic of optimization before this PR:

https://github.com/apache/spark/blob/039fd13eacb1cef835045e3a60cebf958589e1a2/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala#L671-L677

```
        val isLeftSemiWithMatch =
          joinType == LeftSemi && joinSide == LeftSide && iteratorNotEmpty
        // Add to state store only if both removal predicates do not match,
        // and the row is not matched for left side of left semi join.
        val shouldAddToState =
          !stateKeyWatermarkPredicateFunc(key) && !stateValueWatermarkPredicateFunc(thisRow) &&
          !isLeftSemiWithMatch 
```

The criteria of `both removal predicates do not match` means the input is going to be evicted in this batch. I'm not sure about the coverage of this optimization, but there are two major issues with the above optimization: 

1) missing to add the input to state store in left side prevents the input on the right side to match with "that" input. Even though the input is going to be evicted in this batch, there could be still inputs on the right side in this batch which can match with that input.

2) missing to add the input to state store prevents that input to produce unmatched (null-outer) output, as we produce unmatched output during the eviction of state.

Worth noting that `state watermark != watermark for eviction` and eviction we mentioned in above is based on "state watermark". state watermark could be either 1) equal or earlier than watermark for eviction or 2) "later" than watermark for eviction.

### Does this PR introduce _any_ user-facing change?

Yes, there are correctness issues among stream-stream join, especially when the output of the stateful operator is provided as input of stream-stream join. The correctness issue is fixed with the PR.

### How was this patch tested?

New UTs.

### Was this patch authored or co-authored using generative AI tooling?

No.

HeartSaVioR committed:
[SPARK-49829][SS] Revise the optimization on adding input to state store in stream-stream join (correctness fix)

HeartSaVioR committed:
silly fix

HeartSaVioR committed:
scalastyle fix

neilramaswamy commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:
```diff
@@ -668,13 +668,37 @@ case class StreamingSymmetricHashJoinExec(
-          joinType == LeftSemi && joinSide == LeftSide && iteratorNotEmpty
-        // Add to state store only if both removal predicates do not match,
-        // and the row is not matched for left side of left semi join.
-        val shouldAddToState =
-          !stateKeyWatermarkPredicateFunc(key) && !stateValueWatermarkPredicateFunc(thisRow) &&
-          !isLeftSemiWithMatch
```
I imagine that you considered constructing these predicates (e.g. `stateKeyWatermarkPredicateFunc` and `stateValueWatermarkPredicateFunc`) based off the watermark for late events, not the watermark for eviction. I spent some time working this out myself, and I know the subtle reason why this won't work, but I wanted to verify that you also considered this.

HeartSaVioR replied:
It's neither watermark for late event and watermark for eviction - stream-stream join has its own "state watermark", which is going to be an "output watermark" for stream-stream join. The predicate is relying on state watermark.

neilramaswamy replied:
Right, but I had initially wondered why we can't construct the state watermark predicates using the watermark for late events.

For example, with the join predicate `L > R + 10`, we construct the state watermark to be `L <= watermark_for_eviction(R) + 10`, and I had initially thought that the fix for this correctness bug wasn't `L <= watermark_for_late_events(R) + 10`.

HeartSaVioR replied:
Yes, just simply saying, the reason we break down watermark for late event and eviction applies to stream-stream join, except the fact we'll need to buffer event longer than watermark based on join condition, hence necessity of state watermark (and it's related to eviction).

dongjoon-hyun reviewed (commented):
Thank you, @HeartSaVioR .

Is this correctness issue introduced via SPARK-32862 at Apache Spark 3.1.0?
- https://github.com/apache/spark/pull/30076

cc @xuanyuanking,  @viirya too from #30076

neilramaswamy commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:
```diff
@@ -668,13 +668,37 @@ case class StreamingSymmetricHashJoinExec(
+        //   and the join type is left semi.
+        //   For other cases, the input should be added, including the case it's going to be evicted
+        //   in this batch. It hasn't yet evaluated with inputs from right side "for this batch".
+        //   Learn about how the each side figures out the matches from other side.
```
I don't understand what this line means. Can you clarify?

HeartSaVioR replied:
We do not build a hashmap separately to match both sides and completely rely on state store. The way each side figures out the match is, looking into state store of other side. And there is a sequence of doing this, left side, and then right side.

That said, when the operator seeks for match from left side, right side is yet to be handled, hence left side can only see the right side for "prior batches". We take care of match in current batch from right side, assuming that we put the input of left side into state store. If we skip adding the input into state store for left side, we are missing possible matches rows (correctness issue).

HeartSaVioR replied:
If the classdoc covers this I should just say read through classdoc. Do you get how it works from reading through classdoc?

neilramaswamy replied:
I understand the semantics, but I just didn't understand what your comment meant by "Learn about how...". Your latest commit addresses my concern.

neilramaswamy commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:
```diff
@@ -668,13 +668,37 @@ case class StreamingSymmetricHashJoinExec(
+            !stateKeyWatermarkPredicateFunc(key) && !stateValueWatermarkPredicateFunc(thisRow)
+
+          isNotEvictingInThisBatch ||
+            // if the input is producing "unmatched row" in this batch
+            (
+              (joinType == RightOuter && !iteratorNotEmpty) ||
+                (joinType == FullOuter && !iteratorNotEmpty)
```
In the case of a left anti join, isn't it also possible that we would evict this record (and `!iteratorNotEmpty`), so we need to add it to state to produce it?

HeartSaVioR replied:
We don't support left anti join in streaming. supported: inner, left/right outer, full outer, left semi. Please check StreamingSymmetricHashJoinExec.

neilramaswamy commented:
@dongjoon-hyun, this issue has been around since the time that we added multiple stateful operators to Structured Streaming. The Left-semi join logic that you linked previously is correct, and Jungtaek preserves that behavior in this PR.

dongjoon-hyun commented:
To @neilramaswamy , could you provide the affected version list? Actually, I'm tracking it to update SPARK-49829 Affected Version field correctly. Currently it claims `3.5.0+` as the affected version. To be clear, which Spark version (or JIRA ID) do you mean exactly?

> @dongjoon-hyun, this issue has been around since the time that we added multiple stateful operators to Structured Streaming. The Left-semi join logic that you linked previously is correct, and Jungtaek preserves that behavior in this PR.

HeartSaVioR commented:
https://issues.apache.org/jira/browse/SPARK-40925
It was 3.4.0, sorry I forgot that there was a long time between we fixed a bug to support multiple stateful operators partially and we fixed the watermark mechanism to support stream-stream join followed by stateful operator. This is about former.

dongjoon-hyun commented:
Thank you, @HeartSaVioR .

HeartSaVioR committed:
reword

neilramaswamy commented:
Fix makes sense, will thoroughly review tests shortly.

xuanyuanking commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala:
```diff
@@ -878,6 +878,60 @@ class MultiStatefulOperatorsSuite
     testOutputWatermarkInJoin(join3, input1, -40L * 1000 - 1)
   }
 
+  // NOTE: This is the revise of the reproducer in SPARK-45637. CREDIT goes to @andrezjzera.
```
IIRC, if we want to add credit information, we can amend the PR commits to include a co-author. This way, we don’t need to add the credit information in the code comments.

HeartSaVioR replied:
Though the approach of "co-authorship" is not fine-grained one e.g. the credit is only for test, I can do that instead if it's more comfortable to people. Also I don't see we give CREDIT like this way, so probably the preference is what you proposed.

xuanyuanking commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:
```diff
@@ -1966,4 +2046,130 @@ class StreamingLeftSemiJoinSuite extends StreamingJoinSuite {
+  // scalastyle:off line.size.limit
+  // DISCLAIM: This is a revision of below test, which was a part of report in the dev mailing
+  // list. CREDIT goes to @andrezjzera.
+  // https://github.com/andrzejzera/spark-bugs/blob/abae7a3839326a8eafc7516a51aca5e0c79282a6/spark-3.5/src/test/scala/OuterJoinTest.scala#L86C3-L167C4
```
ditto

HeartSaVioR commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:
```diff
@@ -668,13 +668,38 @@ case class StreamingSymmetricHashJoinExec(
+          isNotEvictingInThisBatch ||
+            // if the input is producing "unmatched row" in this batch
+            (
+              (joinType == RightOuter && !iteratorNotEmpty) ||
```
No, the change is following:

1. For left side, we store the new input row into state store regardless of whether the row will be evicted in this batch. This is required because it is still yet to be checked for match with new input rows on the right side "in this batch".
2. For right side, we don't strictly need to store all new input row into state store, but we still need to store them if they are producing "unmatched" output (right/full outer), because we are relying on state eviction to produce unmatched output.

andrzejzera committed:
Give CREDIT to Andrzej Zera for reproducers

HeartSaVioR commented:
@xuanyuanking @neilramaswamy Would you mind continue reviewing? Thanks in advance!

neilramaswamy commented:
One thing I noticed from the unit-tests is that the join-predicate is always exactly an equi-join (not time-interval), and the watermark duration is always 0 units. So this made me wonder—are we missing any coverage in the tests?

Specifically, it seems like this issue arrises because the early-state exclusion logic misses the case where there are some events that are _not_ late, but _are_ evictable by the state watermark predicates. So do there exists cases even without the late watermark being different from the eviction watermark where we might have this bug?

Here's a case I can think of. Suppose the late/eviction watermarks are the same, at something like `t`. Now, suppose that the state watermark is something like `t + k`, maybe because the user has the predicate `L.time > R.time + k`, and if the watermark is `R.time`, then the left eviction predicate will be `L.time <= t + k`. Then, anything greater than `t` will not be dropped due to lateness, and anything less than `t + k` will be early-excluded from the state store. This would make such a record not be added to the state store, and so for left outer/full outer we would not emit the null result properly.

Is my understanding correct?

HeartSaVioR commented:
@neilramaswamy 
Yeah I have to call it out as "state watermark" rather than "watermark for eviction". I originally thought `state watermark <= watermark for eviction`, but you've provided the case of `state watermark > watermark for eviction` which seems to also make sense. 

The case is still handled in this PR, but good to update the PR explanation to replace "watermark for eviction" to "state watermark". Thanks for the input!

HeartSaVioR committed:
empty commit

HeartSaVioR commented:
Build failure came from Spark connect which is not relevant:

```
org.apache.spark.sql.connect.execution.ReattachableExecuteSuite
```

HeartSaVioR commented:
cc. @viirya @brkyvz Hopefully requesting review. Thanks in advance.

neilramaswamy commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:
```diff
@@ -1966,4 +2046,125 @@ class StreamingLeftSemiJoinSuite extends StreamingJoinSuite {
     )
   }
+
+  test("SPARK-49829 two chained stream-stream left outer joins among three input streams") {
```
Can you also comment on this test which record was previously being dropped?

HeartSaVioR replied:
I'll do. Thanks.

HeartSaVioR commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/MultiStatefulOperatorsSuite.scala:
```diff
@@ -878,6 +878,59 @@ class MultiStatefulOperatorsSuite
+    // Inputs are produced into stream-stream join when the time windows are completed, meaning
+    // they will be evicted in this batch for stream-stream join as well. (NOTE: join condition
+    // does not delay the state watermark in stream-stream join).
+    // Before SPARK-49829, left side does not add the input to state store if it's going to evict
+    // in this batch, which breaks the match between input from left side and input from right
+    // side "for this batch".
```
I forgot it - I'll run the test with master and update the code comment.

HeartSaVioR commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:
```diff
@@ -1559,6 +1581,64 @@ class StreamingOuterJoinSuite extends StreamingJoinSuite {
+          // WM: late record = 0, eviction = 2
+          // Before Spark introduces multiple stateful operator, WM for late record was same as
+          // WM for eviction, hence ("d", 1) was treated as late record.
+          // With the multiple state operator, ("d", 1) is added in batch 1 but also evicted in
+          // batch 1. Before SPARK-49829, this wasn't producing unmatched row, and it is fixed.
```
Yes, do you want me to update the code comment? I assume the people who will read the test must be aware of essential behavior of stream-stream join, but probably also good to elaborate as stream-stream join is not very easy operator to understand.

neilramaswamy replied:
Yeah, probably good thing to mention state watermark. In the new test that I will add, I will mention the state watermark. Should be simple here as watermark = 0 and there's no time interval condition. It's just the watermark for eviction.

HeartSaVioR committed:
reflect review comments

neilramaswamy commented:
Just pointing this out for completeness: another way to fix this issue would be to make sure that the state watermark is never larger than the watermark for late events. You can do this by computing the real state watermark to be the minimum of the computed state watermark and the watermark for late events.

Not saying that you should do this, but it would fix the issue too.

HeartSaVioR commented:
The overall direction of watermark is to advance as fast as we see safe and not break the simplicity of current watermark model (there might be trade-off).

I might not put the design discussion into JIRA ticket, but I got an input internally when I designed supporting multiple stateful operators - why not just advance watermark based on state watermark e.g. based on completed windows for window aggregation. This technically delays the advance of watermark by one batch "per operator", due to the mechanism of how we calculate and propagate watermark (at the planning rather than within microbatch). So we rejected it and tolerate some tricky situation like this.

That said, the way we do is by design/intention. If you see the feedback from @andrzejzera who reported the correctness issue, he even said it's uneasy to intuitively follow the behavior because we delay producing output than it is theoretically possible to.
https://lists.apache.org/thread/ysxmtqc1kycthnk0wjmts9sztkt1ofp2

So further delaying to produce output does not sound to me as an option.

neilramaswamy reviewed (approved):
Apologies for the long review time, but I made sure to give a very thorough review. I'm glad we fully understand the situation now, i.e. records falling between the late eviction watermark and state watermark were dropped. Strong LGTM.

HeartSaVioR committed:
reflect feedback

HeartSaVioR commented:
cc. @viirya @brkyvz Friendly reminder.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingJoinSuite.scala:
```diff
@@ -1966,4 +2047,128 @@ class StreamingLeftSemiJoinSuite extends StreamingJoinSuite {
+        Row(Timestamp.valueOf("2024-02-10 10:23:00"), 4, null, null),
+        Row(Timestamp.valueOf("2024-02-10 10:24:00"), 5, 5, 5),
+      )
+       */
```
nit, indentation?

HeartSaVioR replied:
It's a correct indentation; same column for both `*`. 

Spark Scala style guide doesn't have a guide for multiline code comment except the "doc" (class/method/etc), but this is the guide for class/method/etc doc:

```
/**
 * This is correct multi-line JavaDoc comment. And
 * this is my second line, and if I keep typing, this would be
 * my third line.
 */
```

and every starting `*` is placed at the same column.

Please let me know if I'm missing somthing. Let's not block the PR for this - if I'm missing something, I can address it as post-review comment.

dongjoon-hyun reviewed (approved):
Thank you for adding comprehensive test cases together, @HeartSaVioR .

If you don't mind, could you remove `(correctness fix)` from the PR title?

+1 from my side (except two nits; the above PR title and one indentation comment).

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamingSymmetricHashJoinExec.scala:
```diff
@@ -668,13 +668,38 @@ case class StreamingSymmetricHashJoinExec(
+          val isNotEvictingInThisBatch =
+            !stateKeyWatermarkPredicateFunc(key) && !stateValueWatermarkPredicateFunc(thisRow)
+
+          isNotEvictingInThisBatch ||
```
Hm, why `isNotEvictingInThisBatch` is limited to right side join instead of both before? If the reason that it is removed from left side is because 

> missing to add the input to state store in left side prevents the input on the right side to match with "that" input. Even though the input is going to be evicted in this batch, there could be still inputs on the right side in this batch which can match with that input.

Isn't the reason also applied for right side too?

viirya replied:
Oh, nvm, rightSideJoiner already joined new right input with new left input which is just stored (that is wrongly NOT stored previously).

HeartSaVioR replied:
Yes, when evaluating right side, left side of the current batch is expected to be in state store.

HeartSaVioR commented:
I just updated the PR title to change `revise` to `fix the bug` so that it represents we fix the bug. I had to explicitly mention about correctness fix because revising optimization can be considered in two different ways, 1) better optimization 2) fixing issue on the optimization. I'll be direct on the intention for fixing the bug.

HeartSaVioR commented:
Thanks all for the review! Merging to master/3.5/3.4.

HeartSaVioR commented:
(porting to 3.5/3.4 in progress...)

HeartSaVioR commented:
Done. I have only run maven build after cherry-pick; I'll follow the Github Action for these commits to check the test is not failing.
